sql/analyzer/indexed_joins.go

```diff
@@ -651,13 +655,172 @@ func addHashJoins(m *memo.Memo) error {
 	})
 }
 
+type rangeFilter struct {
+	value, min, max                        *memo.ExprGroup
+	closedOnLowerBound, closedOnUpperBound bool
+}
+
+func getRangeFilters(filters []memo.ScalarExpr) (ranges []rangeFilter) {
+	type candidateMap struct {
+		group    *memo.ExprGroup
+		isClosed bool
+	}
+	lowerToUpper := make(map[uint64][]candidateMap)
+	upperToLower := make(map[uint64][]candidateMap)
+
+	findUpperBounds := func(value, min *memo.ExprGroup, closedOnLowerBound bool) {
+		for _, max := range lowerToUpper[memo.InternExpr(value.Scalar)] {
+			ranges = append(ranges, rangeFilter{
+				value:              value,
+				min:                min,
+				max:                max.group,
+				closedOnLowerBound: closedOnLowerBound,
+				closedOnUpperBound: max.isClosed})
+		}
+	}
+
+	findLowerBounds := func(value, max *memo.ExprGroup, closedOnUpperBound bool) {
+		for _, min := range upperToLower[memo.InternExpr(value.Scalar)] {
+			ranges = append(ranges, rangeFilter{
+				value:              value,
+				min:                min.group,
+				max:                max,
+				closedOnLowerBound: min.isClosed,
+				closedOnUpperBound: closedOnUpperBound})
+		}
+	}
+
+	addBounds := func(lower, upper *memo.ExprGroup, isClosed bool) {
+		lowerIntern := memo.InternExpr(lower.Scalar)
+		lowerToUpper[lowerIntern] = append(lowerToUpper[lowerIntern], candidateMap{
+			group:    upper,
+			isClosed: isClosed,
+		})
+		upperIntern := memo.InternExpr(upper.Scalar)
+		upperToLower[upperIntern] = append(upperToLower[upperIntern], candidateMap{
+			group:    lower,
+			isClosed: isClosed,
+		})
+	}
+
+	for _, filter := range filters {
+		switch f := filter.(type) {
+		case *memo.Between:
+			ranges = append(ranges, rangeFilter{f.Value, f.Min, f.Max, true, true})
+		case *memo.Gt:
+			findUpperBounds(f.Left, f.Right, false)
+			findLowerBounds(f.Right, f.Left, false)
+			addBounds(f.Right, f.Left, false)
+		case *memo.Geq:
+			findUpperBounds(f.Left, f.Right, true)
+			findLowerBounds(f.Right, f.Left, true)
+			addBounds(f.Right, f.Left, true)
+		case *memo.Lt:
+			findLowerBounds(f.Left, f.Right, false)
+			findUpperBounds(f.Right, f.Left, false)
+			addBounds(f.Left, f.Right, false)
+		case *memo.Leq:
+			findLowerBounds(f.Left, f.Right, true)
+			findUpperBounds(f.Right, f.Left, true)
+			addBounds(f.Left, f.Right, false)
+		}
+	}
+	return ranges
+}
+
+func addSlidingRangeJoin(m *memo.Memo) error {
```

Comment:
docstring pls. For exploration rules in particular, giving a generalization of (pattern matched) => (rearrangement) and/or a toy SQL example can be helpful

Reply:
Done.